src/subtype.c

```diff
@@ -367,6 +367,17 @@ static int in_union(jl_value_t *u, jl_value_t *x) JL_NOTSAFEPOINT
     return in_union(((jl_uniontype_t*)u)->a, x) || in_union(((jl_uniontype_t*)u)->b, x);
 }
 
+static int obviously_in_union(jl_value_t *u, jl_value_t *x) JL_NOTSAFEPOINT
+{
+    if (jl_is_uniontype(x))
+        return obviously_in_union(u, ((jl_uniontype_t*)x)->a) &&
+               obviously_in_union(u, ((jl_uniontype_t*)x)->b);
```

Comment:
this seems almost certain never to be true (since if a == u and b == u, then a == b, and the union would be elided on construction)?

Reply:
In `simple_join` we use `jl_new_struct(jl_uniontype_type, a, b);` to create `Union`, which has no subtype like `jl_type_union`.
A simple example why we need this is `A = Union{Int, Int8}` and `B = Union{a=Union{Int, Float64}, b=Int8}`.
If we dont split A into `Int` and `Int8`, `obviously_egal` would fail and return false.
Of course this is not sufficient, as I still see `Union{Val{R}, Val{R}, Int, Int8}` when we join `Union{Val{R}, Int}`and `Union{Val{R}, Int8}`.

Reply:
Ah right. I misinterpreted this. On re-reading it is correct